src/contrib/cluster/Akka.Cluster.Sharding.Tests.MultiNode/AsyncWriteProxyEx.cs

```diff
@@ -0,0 +1,461 @@
+        /// </summary>
+        public IStash Stash { get; set; }
+
+        // sent to self only
+        /// <summary>
+        /// TBD
+        /// </summary>
+        public class InitTimeout
+        {
+            private InitTimeout() { }
+            private static readonly InitTimeout _instance = new InitTimeout();
+
+            /// <summary>
+            /// TBD
+            /// </summary>
+            public static InitTimeout Instance
+            {
+                get
+                {
+                    return _instance;
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// TBD
+    /// </summary>
+    internal class ReplayMediator : ActorBase
+    {
+        private readonly Action<IPersistentRepresentation> _replayCallback;
+        private readonly TaskCompletionSource<object> _replayCompletionPromise;
+        private readonly TimeSpan _replayTimeout;
+
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="replayCallback">TBD</param>
+        /// <param name="replayCompletionPromise">TBD</param>
+        /// <param name="replayTimeout">TBD</param>
+        public ReplayMediator(Action<IPersistentRepresentation> replayCallback, TaskCompletionSource<object> replayCompletionPromise, TimeSpan replayTimeout)
+        {
+            _replayCallback = replayCallback;
+            _replayCompletionPromise = replayCompletionPromise;
+            _replayTimeout = replayTimeout;
+
+            Context.SetReceiveTimeout(replayTimeout);
+        }
+
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="message">TBD</param>
+        /// <exception cref="AsyncReplayTimeoutException">
+        /// This exception is thrown when the replay timed out due to inactivity.
+        /// </exception>
+        /// <returns>TBD</returns>
+        protected override bool Receive(object message)
+        {
+            if (message is ReplayedMessage rm)
+            {
+                //rm.Persistent
+                _replayCallback(rm.Persistent);
+            }
+            else if (message is RecoverySuccess)
+            {
+                _replayCompletionPromise.SetResult(new object());
+                Context.Stop(Self);
+            }
+            else if (message is ReplayMessagesFailure)
+            {
+                var failure = message as ReplayMessagesFailure;
+                _replayCompletionPromise.SetException(failure.Cause);
+                Context.Stop(Self);
+            }
+            else if (message is ReceiveTimeout)
+            {
+                var timeoutException = new AsyncReplayTimeoutException($"Replay timed out after {_replayTimeout.TotalSeconds}s of inactivity");
+                _replayCompletionPromise.SetException(timeoutException);
+                Context.Stop(Self);
+            }
+            else return false;
+            return true;
+        }
+    }
+
+    internal static class FuturesEx
```

Comment:
Enhanced version of Ask, which provide the real sender address. Needed this since persistence is using Sender from the commands